File: sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/measure/Measure.java
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.api.batch.sensor.measure;

import java.io.Serializable;
import org.sonar.api.batch.fs.InputComponent;
import org.sonar.api.batch.measure.Metric;

/**
 * Measure on File, Directory or Project.
 * Should not be implemented by client.
 * @since 5.1
 */
public interface Measure<G extends Serializable> {

  /**
   * The {@link InputComponent} this measure belongs to.
   */
  InputComponent inputComponent();

  /**
   * The metric this measure belong to.
   */
  Metric<G> metric();

  /**
   * Value of the measure.
   */
  G value();

}

File: server/sonar-web/src/main/js/components/measure/Measure.js
/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import Rating from '../ui/Rating';
import Level from '../ui/Level';
import Tooltips from '../controls/Tooltip';
import { formatMeasure, isDiffMetric } from '../../helpers/measures';
import { formatLeak, getRatingTooltip } from './utils';
import type { MeasureEnhanced } from './types';

type Props = {
  className?: string,
  measure: MeasureEnhanced,
  decimals?: ?number
};

export default class Measure extends React.PureComponent {
  props: Props;

  renderRating() {
    const { measure } = this.props;
    const metric = measure.metric;
    const value = isDiffMetric(metric.key) ? measure.leak : measure.value;
    const tooltip = getRatingTooltip(metric.key, value);
    const rating = <Rating value={value} />;

    if (tooltip) {
      return (
        <Tooltips overlay={tooltip}>
          <span className={this.props.className}>
            {rating}
          </span>
        </Tooltips>
      );
    }

    return rating;
  }

  render() {
    const { className, decimals, measure } = this.props;
    const metric = measure.metric;

    if (metric.type === 'RATING') {
      return this.renderRating();
    }

    if (metric.type === 'LEVEL') {
      return <Level className={className} level={measure.value} />;
    }

    const formattedValue = isDiffMetric(metric.key)
      ? formatLeak(measure.leak, metric, { decimals })
      : formatMeasure(measure.value, metric.type, { decimals });
    return (
      <span className={className}>
        {formattedValue != null ? formattedValue : '–'}
      </span>
    );
  }
}
